completely revamp config

mixoteric opened the pull request:
# Wikia/config Complete Revamp

Sister Wikia/config PR: https://github.com/Wikia/config/pull/2372

## Stage One

First, I needed to reverse engineer the original vanilla MW behavior with regard to `DefaultSettings.php` and `LocalSettings.php` inclusion.  I then reverse engineered Wikia/app behavior and brought it back to the original shape.  At this point both `app/includes/DefaultSettings.php` and `app/includes/LocalSettings.php` are included in the original order.  `app/includes/wikia/DefaultSettings.php` is skipped at this point which breaks the application.  The problem will be addressed in the next stage.

**Commits:** https://github.com/Wikia/app/pull/14584/commits/41f32e2c8e3d8c5568aced97b3b18778e873880d https://github.com/Wikia/app/pull/14584/commits/9819761558bcfea7e99f710d821b9f9d9baf8f2d https://github.com/Wikia/app/pull/14584/commits/1593222dec376aa480afe5300cd374607b1d7dfb https://github.com/Wikia/app/pull/14584/commits/0c51e864c9252f6b938fb269064b7edfaffe1f8c https://github.com/Wikia/app/pull/14584/commits/5416e6d96fb386b5df47dd369898784a96bc0c50 https://github.com/Wikia/app/pull/14584/commits/c84ae226281c55b435c39f45f2c59f5968ce08b7

## Stage Two

At this stage I needed to strip down Wikia/config as much as possible.  I focused on the code executed before WikiFactory configuration is loaded.  I moved some insensitive data to `LocalSettings.php` and re-applied FANDOM-specific values in `includes/wikia/DefaultSettings.php` to override defaults from `includes/DefaultSettings.php`.  At this point Wikia/config and Wikia/app branches should work together and are expected to produce the same configuration as they would with `dev` branches.

**Commits:** https://github.com/Wikia/app/pull/14584/commits/525db43e99fdea1b674ed2d1fbd4c1f581bb6298 https://github.com/Wikia/app/pull/14584/commits/7aa2a79e5bc05f89c2963f7aace65a1da5d8722b https://github.com/Wikia/app/pull/14584/commits/2d9793d65e1030f9fb21c788006b685db39e85a2 https://github.com/Wikia/app/pull/14584/commits/e9dc1f6b2ffb8e74de019134ff330e7d25f9a5ab https://github.com/Wikia/app/pull/14584/commits/ed0ff38208543a00d7d60147d0d508a639843279 https://github.com/Wikia/app/pull/14584/commits/d2360d54ba1e982a8066b274b3abaa43d63d453f https://github.com/Wikia/app/pull/14584/commits/dc849cf87295168b637131a9d1642a8e0ce2b193 https://github.com/Wikia/app/pull/14584/commits/80ef9dd029ccfc2f2aa495f8a8a225549ab7ceb6 https://github.com/Wikia/app/pull/14584/commits/593d2fed54df52d423832320658039cfa547c89b https://github.com/Wikia/app/pull/14584/commits/4e2a12b2f42e453d7d5132f4615e712aaa07505f

## Stage Three

Clean up `CommonSettings.php`.

Commit:
remove unneeded $IP declaration

$IP is set properly by core MediaWiki code in:

* includes/WebStart.php
* maintenance/Maintenance.php

Commit:
update include_path once $IP is declared

* removed from WikiFactoryLoader
* placed in includes/WebStart.php and maintenance/Maintenance.php

This is a little bit of overkill as the convention is to use
absolute paths and $IP everywhere.  Yet, still a good idea
to add most common include paths to the list.

Commit:
remove unneeded includes/Defines.php import

includes/Defines.php is imported earlier in core MW execution flow.

Commit:
include DefaultSettings.php as core MW does it

Commit:
import DefaultSettings.php from vanilla MW 1.19.24

All FANDOM-specific settings that are different than defaults will be
set in FANDOM's custom configuration.  This file really should not be
edited.

mszabo-wikia reviewed (dismissed):
Backport of vanilla DefaultSettings.php does not account for any changes backported from newer MW security releases. I see at least two - `wgBlockCIDRLimit` and `wgMinUploadChunkSize`. Let's audit the file history so that we do not miss anything.

ghost commented:
@TK-999 — that's my plan for Stage Two.  My goal was to go back to vanilla 1.19.24 and progress from there with proper design and implementation, leaving the original `DefaultSettings.php` untouched.  I will also create a testing script that will compare `Wikia/app@dev` + `Wikia/config@dev` final configuration for top wikis with config revamp branches to make sure we have identical results.

Commit:
do not include DefaultSettings.php in WikiFactory

Commit:
include Hooks.php as core MW does it

Commit:
load GlobalSettings.php earlier

In core MediaWiki execution flow, includes/GlobalFunctions.php are
loaded in includes/Setup.php which would be too late for WikiFactory.

WikiFactory injects configuration in LocalSettings.php, whereas
includes/Setup.php is loaded afterwards.

I have reviewed includes/GlobalFunctions.php and considered it safe to
be loaded earlier in the flow. Loading it right after class autoloader
seems convenient.

Commit:
load FANDOM-specific stuff in LocalSettings.php

I have moved the following FANDOM-specific files to the top
of LocalSettings.php so that the code in the is available before
WikiFactory starts injecting configuration.

* includes/wikia/Defines.php
* includes/wikia/DefaultSettings.php
* includes/wikia/GlobalFunctions.php

They are now included after corresponding core MediaWiki files:

* includes/Defines.php
* includes/DefaultSettings.php
* includes/GlobalFunctions.php

Commit:
do not explicitly require includes/Exception.php

includes/Exception.php only contains classes which can be lazy-loaded
by class autoloader.  Explicit import is not required.

Commit:
do not explicitly require includes/db/Database.php

includes/db/Database.php only contains classes which can be lazy-loaded
by class autoloader.  Explicit import is not required.

Commit:
load WikiFactory in includes/wikia/DefaultSettings.php

Commit:
remove unneeded wfUnserializeHandler duplicate

Commit:
load WikiFactoryLoader with class autoloader

Commit:
import env/dc constants from Wikia/config

This is the final stop for them, before they will be remove completely.

Commit:
strip down the config, pt. 1

Commit:
reapply FANDOM defaults

Commit:
import non-sensitive settings from config

Commit:
normalize configuration variables

Commit:
config variables normalization
